### PR TITLE
Firefox driver support

### PR DESCRIPTION
**Pull request by oneumyvakin**
1. FireFox driver doesn't search for firefox binary so user have to specify "binary" capability.
2. Unfortunately driver doesn't support "/status" command, I've tried to do something smart with using bus.Connect to check status but it starts crashing, so I've stopped on "response.StatusCode == 404".

### CODE REVIEW

**Comment by sclevine** [comment]
Hi @oneumyvakin,

Sorry for the delay. I'd prefer not to accept a 404 from the `/status` endpoint in all cases. Is there some other endpoint in Firefox Driver you could check if `/status` returns a 404?

**Comment by oneumyvakin** [comment]
I've four options:
1. In case of 404, force to create session, delete session, hardcode sleep to wait for browser closing.
2. In case of 404, parse response for {"error":"unknown command","message":"GET /status did not match a known command"}. 
3. In case of 404, create GET request for default marionette port 127.0.0.1:2828, parse response for {"applicationType":"gecko","marionetteProtocol":3}
4. Wait until /status will be implemented =)

**Comment by sclevine** [comment]
Any update on a `/status` endpoint?

**Comment by oneumyvakin** [comment]
@sclevine Yes, thanks. Finally it works on latest geckodriver 0.17. 
There is small change to unwrap session ID. I've checked FirefoxDriver on latest FF as well as ChromeDriver to make sure I've not broke something.

File: agouti.go [hunk 1]
```diff
@@ -104,3 +104,42 @@ func SauceLabs(name, platform, browser, version, username, accessKey string, opt
 	capabilities["name"] = name
 	return NewPage(url, append([]Option{Desired(capabilities)}, options...)...)
 }
+
+// Example:
```
Comment (sclevine) [inline agouti.go]: Example is not necessary here.

File: agouti.go [hunk 2]
```diff
@@ -104,3 +104,42 @@ func SauceLabs(name, platform, browser, version, username, accessKey string, opt
...
+//
+// 	capabilities := agouti.NewCapabilities()
+//	capabilities[""] = ""
```
Comment (sclevine) [inline agouti.go]: Is this required for some reason?

File: api/internal/bus/connect.go [hunk 3]
```diff
@@ -58,19 +58,35 @@ func openSession(url string, body io.Reader, httpClient *http.Client) (sessionID
...
-	var sessionResponse struct{ SessionID string }
+	var sessionResponse struct {
+		SessionID string
+	}
 	responseBody, err := ioutil.ReadAll(response.Body)
 	if err != nil {
 		return "", err
 	}
 
-	if err := json.Unmarshal(responseBody, &sessionResponse); err != nil {
```
Comment (sclevine) [inline api/internal/bus/connect.go]: Revert this.

File: api/internal/bus/connect.go [hunk 4]
```diff
@@ -58,19 +58,35 @@ func openSession(url string, body io.Reader, httpClient *http.Client) (sessionID
...
 
-	if sessionResponse.SessionID == "" {
-		return "", errors.New("failed to retrieve a session ID")
+	if sessionResponse.SessionID != "" {
+		return sessionResponse.SessionID, nil
 	}
 
-	return sessionResponse.SessionID, nil
-}
+	var firefoxResponse struct {
```
Comment (sclevine) [inline api/internal/bus/connect.go]: Struct name shouldn't be Firefox-specific.

File: api/internal/bus/connect.go [hunk 5]
```diff
@@ -58,19 +58,35 @@ func openSession(url string, body io.Reader, httpClient *http.Client) (sessionID
...
+	if err = json.Unmarshal(responseBody, &firefoxResponse); err != nil {
+		return "", err
+	}
+
+	if firefoxResponse.Value.SessionID != "" {
+		return firefoxResponse.Value.SessionID, nil
+	}
+
+	return "", errors.New("failed to retrieve a session ID")
+}
```
Comment (sclevine) [inline api/internal/bus/connect.go]: Restore newline.

**Comment by sclevine** [comment]
Closing in favor of #150.

**Pull request closed without merging**